packages/core/src/lib/base_protocol.ts

```diff
@@ -28,13 +29,26 @@ export class BaseProtocol {
   }
 
   protected async getPeer(peerId?: PeerId): Promise<Peer> {
-    const { peer } = await selectPeerForProtocol(
-      this.peerStore,
-      [this.multicodec],
-      peerId
-    );
+    let peer = await this._getPeer(peerId);
+    if (!peer) {
+      await waitForRemotePeer(this.multicodec, this.peerStore);
+      peer = (await this._getPeer(peerId)) as Peer;
+    }
     return peer;
   }
+
+  protected async _getPeer(peerId?: PeerId): Promise<Peer | undefined> {
+    try {
+      const { peer } = await selectPeerForProtocol(
+        this.peerStore,
+        [this.multicodec],
+        peerId
+      );
+      return peer;
+    } catch (error) {
```

Comment:
What kind of error can be thrown? `selectPeerForProtocol` shouldn't throw, right?